pallets/creditcoin/src/helpers/external_address.rs

```diff
@@ -1,14 +1,63 @@
 use crate::{Blockchain, ExternalAddress};
 use base58::FromBase58;
+use core::convert::TryFrom;
+use frame_support::BoundedVec;
+use sp_core::ecdsa::Public;
+use sp_io::hashing::keccak_256;
 use sp_io::hashing::sha2_256;
 
+pub fn external_address_generator(
+	blockchain: &Blockchain,
+	reference: &ExternalAddress,
+	pkey: Public,
+) -> Option<ExternalAddress> {
+	match blockchain {
+		Blockchain::Luniverse | Blockchain::Ethereum | Blockchain::Rinkeby
+			if Etherlike::is_address(reference).is_some() =>
+		{
+			Some(Etherlike::from_public(&pkey))
+		},
+		Blockchain::Bitcoin => None,
+		Blockchain::Other(_) => None,
+		_ => None,
+	}
+}
+
+pub trait PublicToAddress {
+	type AddressType;
+	fn is_address(addr: &ExternalAddress) -> Option<Self::AddressType>;
+	fn from_public(pkey: &Public) -> ExternalAddress;
+}
+
+pub struct Etherlike;
+
+impl PublicToAddress for Etherlike {
+	type AddressType = ();
+	fn is_address(addr: &ExternalAddress) -> Option<Self::AddressType> {
```

Comment:
what is the benefit of an `Option` over a `bool` here?

Reply:
You will know if it is a valid blockcahin address plus getting back the address type altogether, you can match on the return type to generate the desired address.

Reply:
It is trivial now that EVMAddress is a struct instead of an enum but, bitcoin and other blockchains will need to match on multiple address protocols if you really want to decouple functions.

Reply:
I think the name `is_address` implies that you are passing it a parameter that is maybe an address. If it is it will return true, if not false. If the goal is to extract the address type then I would expect something like `try_extract_addresss_type`.